src/chsh.1
.\" -*- nroff -*-
.\" Copyright (C) 2002, 2003, 2004 Thorsten Kukuk
.\" Author: Thorsten Kukuk <kukuk@suse.de>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2 as
.\" published by the Free Software Foundation.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software Foundation,
.\" Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.TH chsh 1 "February 2004" "pwdutils"
.SH NAME
chsh \- change login shell
.SH SYNOPSIS
.TP 5
\fBchsh\fR
[\fB-D \fIbinddn\fR] [\fB-P \fIpath\fR] [\fB-s \fIshell\fR] [\fB-l\fR] [\fB-q\fR] [\fB-u\fR] [\fB-v\fR] [\fIuser\fR]
.SH DESCRIPTION
\fBchsh\fR is used to change the user login shell.
A normal user may only change the login shell for their own account,
the super user may change the login shell for any account.
.PP
If a shell is not given on the command line, \fBchsh\fR operates in
an interactive fashion, prompting the user with the current login shell.
Enter the new value to change the field, or leave the line blank to use
the current value. Enter \fInone\fR to remove the current value.
The current value is displayed between a pair of \fB[ ]\fR marks.
.PP
The only restrictions placed on the login shell is that the
command name must be listed in \fI/etc/shells\fR, unless the
invoker is the super-user, and then any value may be added.
An account with a restricted login shell may not change
their login shell.
.PP
This version of \fBchsh\fR is able to change the shell of
local, NIS, NIS+ and LDAP accounts , if the permissions allow
it.
.SH OPTIONS
.TP
.BI "\-D, \-\-binddn" " binddn"
Use the Distinguished Name binddn to bind to the LDAP directory.
The user will be prompted for a password for simple authentication.
.TP
.BI "\-P, \-\-path" " path"
The \fIpasswd\fR file is located below the specified directory
path. \fBchsh\fR will use this files, not \fI/etc/passwd\fR.
This is useful for example on NIS master servers, where you do
not want to give all users in the NIS database automatic access
to your NIS server and the NIS map is build from special files.
.TP
.B "\-s, \-\-shell"
Specify your login shell.
.TP
.B "\-l, \-\-list-shells"
Print the list of shells listed in
.I /etc/shells
and exit.
.TP
.B "\-q, \-\-quite"
Don't be verbose.
.TP
.B "\-u, \-\-usage"
Print a usage message and exit.
.TP
.B "    \-\-help"
Print a more verbose help text and exit.
.TP
.B "-v, \-\-version"
Print version information and exit.
.SH FILES
/etc/passwd \- user account information
.br
/etc/shells \- list of valid login shells
.SH SEE ALSO
.BR chfn (1),
.BR passwd (5),
.BR shells (5)
.SH AUTHOR
Thorsten Kukuk <kukuk@suse.de>


NEWS
pwdutils NEWS -- history of user-visible changes.

Copyright (C) 2002-2012 Thorsten Kukuk

Please send bug reports, questions and suggestions to <kukuk@thkukuk.de>.

Version 3.2.19
* Fix passwd -S

Version 3.2.18
* Fix changing of permissions if uid/gid of user will be modified

Version 3.2.17
* Fix useradd.local

Version 3.2.16
* Make sha512 new default password hash

Version 3.2.15
* newgrp: honour dynamically assigned groups

Version 3.2.14
* pwck: check +/- entries, too
* grpck: check */- entries, too
* login.defs: remove LOGIN_RETRIES
 
Version 3.2.13
* Only copy regular files if usermod moves home directory

Version 3.2.12
* Make sure modified group files are written to disk
* Update translations for sv, fr

Version 3.2.11
* Update translations
* Add translation for nl

Version 3.2.10
* Fix encoding of polish translations
* Fix long options of grpck/pwck
 
Version 3.2.9
* Move UMASK from /etc/login.defs to /etc/default/useradd

Version 3.2.8
* Fix usermod manual page
* Add quiet option to pam_rpasswd.so

Version 3.2.7
* Fix possible endless loop in userdel

Version 3.2.6
* Fix rpasswd protocol error
* Update translations

Version 3.2.5
* Add translations for da
* pwconv: Remove temporary files in case of an error

Version 3.2.4
* Make sure changes are really written to disk

Version 3.2.3
* New translations
* Adjust documentation

Version 3.2.2
* Don't lock system accounts by default

Version 3.2.1
* Adjust for new libxcrypt

Version 3.2.0
* Better support for audit plugin
* --preferred-uid/gid option for useradd and groupadd

Version 3.1.4
* Update polish translation.
* Remove FTMP references.
* Better pwconv support for  NIS entries

Version 3.1.3
* Allow compiling without native language support.
* Update swedish translation.

Version 3.1.2
* Check if crontab is installed before calling it.
* Fix seg.fault in pwconv in error case.

Version 3.1.1
* Adjust PAM config files

Version 3.1.0
* gpasswd: Add -l and -u option to lock/unlock group passwords
* passwd: Add support for BioAPI
* Add LDAP support for RFC2307bis

Version 3.0.7
* passwd.1: correct manual page
* passwd: Allow changing passwords with PAM even if service is unknown

Version 3.0.6
* chfn: Fix syntax check for home phone number

Version 3.0.5
* Fix compilation error of PAM module with new Linux-PAM release
* Fix CHARACTER_CLASS in /etc/login.defs

Version 3.0.4
* Fix pam_rpasswd.so

Version 3.0.3
* usermod: Adjust sp_lstchg if -p option is used
* useradd: Preserve existing home directory
* chage: Fix off-by-one errors in output

Version 3.0.2
* Fix loading of logging plugins
* Unify options (use always --service)
* Add -a option to usermod for shadow compatibility

Version 3.0.1
* usermod: Don't remove old home directory if an error occured

Version 3.0
* Add --stdin option for passwd
* Fix useradd -D option
* Fix PAM/syslog interaction
* Fix lot of ignored failing syscalls
* Compile daemons with -fpie/-pie
* Add support for logging plugins
* Add plugin for LAuS support
* Add plugin for audit support
* Add plugin to log with syslog
* Use libnscd
* Add support for GnuTLS
* groupmod: Add options to remove/add single user from a group
* usermod: Add options to remove/add single user from a list of groups
* rpasswd/rpasswdd: Better SLP support

Version 2.6.4
* rpasswd/pam_rpasswd.so: Fix SLP URL parsing

Version 2.6.3
* newgrp: Add -c option to allow sg emulation

Version 2.6.2
* Fix compile problem if SELinux is disabled
* rpasswdd: Re-register SLP service after timeout

Version 2.6.1
* newgrp: Fix NGROUPS_MAX handling

Version 2.6
* Add -f compat flag to groupadd
* Add UTF8/local locale translation
* Make regex for allowed accounts/groups configurable
* Add optional pam_rpasswd module
* Add SLP support to rpasswd/rpasswdd
* Add pwck
* Add grpck
* Add newgrp
* Add /etc/default/passwd support

Version 2.5.1
* Install all manual pages

Version 2.5
* Adjust for libselinux v1.5
* Add pwconv/pwunconv
* Add grpunconv

Version 2.4
* Fix -G handling of usermod
* Fix LDAP support
* Add gpasswd, groupmod and usermod
* Copy always all extended attributes
* Add SELinux support for passwd, chsh and chfn
* Make sure that group, passwd and shadow are always created with
  correct security attributes
* groupadd -g handling fixed
* vipw/vigr added
* Lot of bug fixes
* Add manual pages
* Add groupadd and groupdel
* useradd and userdel can now create/delete LDAP accounts
* Make compile with OpenLDAP 2.0.x again
* Rewrite ldap functions to meet demands
* chpasswd can now change LDAP passwords
* useradd: Lot of new error checks and bug fixes
* chfn/chsh: Allow user to abort with Ctrl-C during data input
* Add test suite

Version 2.3
* Add chpasswd
* Add expiry
* Fix chage options
* rpasswd client now checks server certificate

Version 2.2
* Add "-D binddn" option, to allow modifying LDAP data with a special
  account.
* passwd and chage: Allow modifying of password expiry informations
  stored in a LDAP database.

Version 2.1
* Add --disable-ldap option to configure
* Fix typos in chage.1 manual page
* Rewrite passwd manual page and explain all options

Version 2.0
* Fix -d and -E option of chage
* Use port 774 as new default
* Make compile with glibc 2.3 and gcc 2.95/3.2/3.3

Version 1.99
* Add chage
* Allow modifying of passwd/shadow file not in /etc (only chage yet)

Version 1.98
* Add manual pages for rpasswd and rpasswdd

Version 1.97
* rpasswd now reads config file with server name and port number
* Security fix for rpasswdd: don't allow remote changing of root
  password without authentification
* Add manual page for rpasswd.conf

Version 1.96
* Make compile with gcc 2.x
* root can now change passwords for other user
* timeout if client does not answer fast enough
* Sent LANG as locale to server

Version 1.95
* Make location of ldap config files compile time configurable

Version 1.94
* Add more sanity checks to prevent file corruption

Version 1.93
* Minor bug fixes to rpasswdd

Version 1.92

src/chsh.c
/* Copyright (C) 2002, 2003, 2004, 2005 Thorsten Kukuk
   Author: Thorsten Kukuk <kukuk@suse.de>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License version 2 as
   published by the Free Software Foundation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <pwd.h>
#include <ctype.h>
#include <errno.h>
#include <stdio.h>
#include <signal.h>
#include <syslog.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <shadow.h>
#ifdef WITH_SELINUX
#include <selinux/selinux.h>
#include <selinux/av_permissions.h>
#endif
#ifdef HAVE_LIBNSCD_H
#include <libnscd.h>
#endif

#include "i18n.h"
#include "public.h"
#include "utf8conv.h"
#include "logindefs.h"
#include "read-files.h"
#include "error_codes.h"

#ifdef USE_LDAP
#include "libldap.h"
#endif

static void
print_usage (FILE *stream, const char *program)
{
  fprintf (stream, _("Usage: %s [-D binddn] [-P path] [-s shell] [-l] [-q]\n            [--help] [--usage] [--version] [user]\n"),
	   program);
}

static void
print_help (const char *program)
{
  print_usage (stdout, program);
  fprintf (stdout, _("%s - change login shell\n\n"), program);

#ifdef USE_LDAP
  fputs (_("  -D binddn      Use dn \"binddn\" to bind to the LDAP directory\n"),
         stdout);
#endif
  fputs (_("  -P path        Search passwd and shadow file in \"path\"\n"),
         stdout);
  fputs (_("  -l             List allowed shells from /etc/shells\n"),
         stdout);
  fputs (_("  -s shell       Use 'shell' as new login shell\n"), stdout);
  if (strcmp (program, "chsh") == 0)
    fputs (_("  --service srv  Use nameservice 'srv'\n"), stdout);
  fputs (_("  -q, --quiet    Don't be verbose\n"), stdout);
  fputs (_("      --help     Give this help list\n"), stdout);
  fputs (_("  -u, --usage    Give a short usage message\n"), stdout);
  fputs (_("  -v, --version  Print program version\n"), stdout);

  if (strcmp (program, "chsh") == 0)
    fputs (_("Valid services are: files, nis, nisplus, ldap\n"),
	   stdout);
}

/* If the given shell appears in /etc/shells, return 1. If not,
   return 0. If the given shell is NULL, /etc/shells is outputted
   to stdout. */
static int
get_shell_list (const char *shell_name)
{
#ifndef HAVE_GETUSERSHELL
  FILE *fp;
  char buf[BUFSIZ];
#else
  char *buf;
#endif
  int found;
  int len;

  found = 0;

#ifdef HAVE_GETUSERSHELL
  setusershell();
#else
  fp = fopen ("/etc/shells", "r");
  if (! fp)
    {
      if (! shell_name)
	printf (_("No known shells.\n"));
      return 1;
    }
#endif

#ifdef HAVE_GETUSERSHELL
  while ((buf = getusershell ()))
#else
  while (fgets (buf, sizeof (buf), fp) != NULL)
#endif
    {
      /* ignore comments */
      if (*buf == '#')
	continue;
      len = strlen (buf);
      /* strip the ending newline */
      if (buf[len - 1] == '\n')
	buf[len - 1] = 0;
      /* check or output the shell */
      if (shell_name)
	{
	  if (! strcmp (shell_name, buf))
	    {
	      found = 1;
	      break;
            }
        }
      else
	printf ("%s\n", buf);
    }
#ifdef HAVE_GETUSERSHELL
  endusershell();
#else
  fclose (fp);
#endif
  return found;
}

/* return 1 if the named shell begins with 'r' or 'R'
   If the first letter of the filename is 'r' or 'R', the shell is
   considered to be restricted. */
static int
restricted_shell (const char *sh)
{
#if 0
  char *cp = Basename((char *) sh);
  return *cp == 'r' || *cp == 'R';
#else
  /* Shells not listed in /etc/shells are considered to be
     restricted.  Changed this to avoid confusion with "rc"
     (the plan9 shell - not restricted despite the name
     starting with 'r'). */
  return !get_shell_list (sh);
#endif
}

/* If the shell is completely invalid, print an error and
   return 1. If root changes the shell, print only a warning.
   Only exception: Invalid characters are always not allowed.  */
static int
check_shell (const char *program, const char *shell)
{
  uid_t uid = getuid ();
  size_t i;
  int c;

  if (*shell != '/')
    {
      fprintf (stderr, _("%s: Shell must be a full path name.\n"), program);
      if (uid)
	return 1;
    }
  if (access (shell, F_OK) < 0)
    {
      fprintf (stderr, _("%s: `%s' does not exist.\n"), program, shell);
      if (uid)
	return 1;
    }
  if (access (shell, X_OK) < 0)
    {
      fprintf (stderr, _("%s: `%s' is not executable.\n"), program, shell);
      if (uid)
	return 1;
    }
  /* keep /etc/passwd clean. */
  for (i = 0; i < strlen (shell); i++)
    {
      c = shell[i];
      if (c == ',' || c == ':' || c == '=' || c == '"' || c == '\n')
	{
	  fprintf (stderr, _("%s: '%c' is not allowed.\n"), program, c);
	  return 1;
        }
      if (iscntrl (c))
	{
	  fprintf (stderr, _("%s: Control characters are not allowed.\n"),
		   program);
	  return 1;
        }
    }
  if (! get_shell_list (shell))
    {
      if (uid == 0)
	printf (_("Warning: \"%s\" is not listed in /etc/shells.\n"), shell);
      else
	{
	  fprintf (stderr, _("%s: \"%s\" is not listed in /etc/shells.\n"),
		   program, shell);
	  fprintf (stderr, _("%s: Use -l option to see list.\n"), program);
	  return 1;
	}
    }
  return 0;
}


int
main (int argc, char *argv[])
{
  uid_t uid = getuid ();
  char *new_shell = NULL;
  int l_flag = 0;
  int silent = 0;
  user_t *pw_data = NULL;
  char *use_service = NULL;
  char *caller_name = NULL;
  char *locale_name;
  const char *program = basename (argv[0]);
#ifdef USE_LDAP
  char *binddn = NULL;
#endif

#ifdef ENABLE_NLS
  setlocale(LC_ALL, "");
  bindtextdomain(PACKAGE, LOCALEDIR);
  textdomain(PACKAGE);
#endif

  openlog (program, LOG_PID, LOG_AUTHPRIV);

  /* Before going any further, raise the ulimit and ignore
     signals.  */
  init_environment ();

  if (strcasecmp (program, "ypchsh") == 0)
    use_service = "nis";
  else if (strcasecmp (program, "chsh") != 0)
    {
      fprintf (stderr, _("%s: Don't know what I should do.\n"), program);
      return 1;
    }

  while (1)
    {
      int c;
      int option_index = 0;
      static struct option long_options[] =
	{
#ifdef USE_LDAP
	  {"binddn",      required_argument, NULL, 'D' },
#endif
	  {"path",        required_argument, NULL, 'P' },
	  {"shell",       required_argument, NULL, 's' },
	  {"list-shells", no_argument,       NULL, 'l' },
	  {"quiet",       no_argument,       NULL, 'q' },
	  {"version",     no_argument,       NULL, 'v' },
	  {"usage",       no_argument,       NULL, 'u' },
	  {"service",     required_argument, NULL, '\254' },
	  {"help",        no_argument,       NULL, '\255' },
	  {NULL,          0,                 NULL, '\0'}
	};

      c = getopt_long (argc, argv, "D:P:s:r:lvuq",
		       long_options, &option_index);
      if (c == (-1))
        break;
      switch (c)
        {

        case 'l':
	  l_flag = 1;
	  break;
        case 's':
	  if (! optarg)
	    {
	      print_usage (stderr, program);
	      return 1;
            }
	  new_shell = strdup (optarg);
	  break;
	case '\254':
	  if (use_service != NULL)
	    {
	      print_usage (stderr, program);
	      return 1;
	    }

	  if (strcasecmp (optarg, "yp") == 0 ||
	      strcasecmp (optarg, "nis") == 0)
	    use_service = "nis";
	  else if (strcasecmp (optarg, "nis+") == 0 ||
		   strcasecmp (optarg, "nisplus") == 0)
	    use_service = "nisplus";
	  else if (strcasecmp (optarg, "files") == 0)
	    use_service = "files";
#ifdef USE_LDAP
	  else if (strcasecmp (optarg, "ldap") == 0)
	    use_service = "ldap";
#endif
	  else
	    {
	      fprintf (stderr, _("Service `%s' not supported.\n"), optarg);
	      print_usage (stderr, program);
	      return 1;
	    }
	  break;
	case 'q':
	  silent = 1;
	  break;
#ifdef USE_LDAP
        case 'D':
          binddn = optarg;
          break;
#endif
        case 'P':
          if (uid != 0)
            {
              fprintf (stderr,
                       _("Only root is allowed to specify another path\n"));
              return E_NOPERM;
            }
          else
            files_etc_dir = strdup (optarg);
          break;
        case '\255':
          print_help (program);
          return 0;
        case 'v':
          print_version (program, "2005");
          return 0;
        case 'u':
          print_usage (stdout, program);
          return 0;
        default:
          print_error (program);
          return 1;
        }
    }

  argc -= optind;
  argv += optind;

  if (argc > 1 || (l_flag && argc > 0))
    {
      fprintf (stderr, _("%s: Too many arguments.\n"), program);
      print_error (program);
      return 1;
    }

  if (l_flag && new_shell)
    {
      fprintf (stderr, _("%s: Too many arguments.\n"), program);
      print_error (program);
      return 1;
    }

  if (l_flag)
    {
      get_shell_list (NULL);
      return 0;
    }
  else
    {
      int buflen = 256;
      char *buffer = alloca (buflen);
      struct passwd resultbuf;
      struct passwd *pw;

      /* Determine our own user name for PAM authentication.  */
      while (getpwuid_r (uid, &resultbuf, buffer, buflen, &pw) != 0
	     && errno == ERANGE)
	{
	  errno = 0;
	  buflen += 256;
	  buffer = alloca (buflen);
	}
      if (!pw)
	{
	  fprintf (stderr, _("%s: Cannot determine your user name.\n"),
		   program);
	  return 1;
	}

      caller_name = strdupa (pw->pw_name);

      /* We change the passwd information for another user, get that
	 data, too.  */
      if (argc == 1)
	{
	  char *user = locale_to_utf8 (argv[0]);

	  while (getpwnam_r (user, &resultbuf, buffer, buflen, &pw) != 0
		 && errno == ERANGE)
	    {
	      errno = 0;
	      buflen += 256;
	      buffer = alloca (buflen);
	    }

	  free (user);
	  if (!pw)
	    {
	      fprintf (stderr, _("%s: Unknown user `%s'.\n"),
		       program, argv[0]);
	      return 1;
	    }
	}

      pw_data = do_getpwnam (pw->pw_name, use_service);
      if (pw_data == NULL || pw_data->service == S_NONE)
	{
	  if (use_service)
	    fprintf (stderr, _("%s: User `%s' is not known to service `%s'.\n"),
		     program, utf8_to_locale (pw->pw_name), use_service);
	  else
	    fprintf (stderr, _("%s: Unknown user `%s'.\n"), program,
		     utf8_to_locale (pw->pw_name));
	  return 1;
	}

      locale_name = utf8_to_locale (pw_data->pw.pw_name);
    }

#ifdef WITH_SELINUX
  if (is_selinux_enabled () > 0)
    {
      if ((uid == 0) &&
          (selinux_check_access (pw_data->pw.pw_name, PASSWD__CHSH) != 0))
        {
          security_context_t user_context;
          if (getprevcon (&user_context) < 0)
            user_context =
              (security_context_t) strdup (_("Unknown user context"));
	  fprintf (stderr,
		  _("%s: %s is not authorized to change the shell of `%s'.\n"),
		   program, user_context, locale_name);
	  if (security_getenforce() > 0)
	    {
	      syslog (LOG_ALERT,
		      "%s is not authorized to change the shell of `%s'",
		      user_context, pw_data->pw.pw_name);
	      freecon (user_context);
	      return E_NOPERM;
	    }
	  else
	    {
	      fprintf (stderr,
		       _("SELinux is in permissive mode, continuing\n"));
	      freecon (user_context);
	    }
        }
    }
#endif


  /* Only root is allowed to change shell for local users. */
  if (uid && uid != pw_data->pw.pw_uid &&
      (pw_data->service == S_LOCAL
#ifdef USE_LDAP
       || (pw_data->service == S_LDAP && binddn == NULL)
#endif
       ))
    {
      syslog (LOG_ERR, "%u cannot change shell for \"%s\"", uid,
	      pw_data->pw.pw_name);
      fprintf (stderr, _("You cannot change the shell for %s.\n"),
	       caller_name);
      free_user_t (pw_data);
      return 1;
    }

  /* Normal user with restricted shell is not allowed to change it. */
  if (uid && restricted_shell (pw_data->pw.pw_shell))
    {
      syslog (LOG_ERR, "User `%s' tries to change a restricted shell",
	      pw_data->pw.pw_name);
      fprintf(stderr, _("You cannot change a restricted shell.\n"));
      free_user_t (pw_data);
      return 1;
    }

  if (!silent)
    printf (_("Changing login shell for %s.\n"), locale_name);

  if (getlogindefs_bool ("CHFN_AUTH", 1) || pw_data->service != S_LOCAL)
    {
#ifdef USE_LDAP
      if (binddn && pw_data->service == S_LDAP)
        {
          /* A user tries to change data stored in a LDAP database and
             knows the Manager dn, now we need the password from him.  */
          ldap_session_t *session = create_ldap_session (LDAP_PATH_CONF);
          char *cp;

          if (session == NULL)
            return E_FAILURE;

          cp = getpass (_("Enter LDAP Password:"));

          pw_data->binddn = strdup (binddn);

          if (open_ldap_session (session) != 0)
            return E_FAILURE;

          if (ldap_authentication (session, NULL, binddn, cp) != 0)
            return E_NOPERM;

          close_ldap_session (session);

          pw_data->oldclearpwd = strdup (cp);
        }
      else
#endif /* USE_LDAP */
	if (do_authentication (program, caller_name, pw_data) != 0)
	  {
	    free_user_t (pw_data);
	    return 1;
	  }
      if (pw_data->service != S_LOCAL && get_old_clear_password (pw_data) != 0)
	return 1;
    }

  if (new_shell == NULL)
    {
      /* Allow user to abort with Ctrl-C here.  */
      signal (SIGINT, SIG_DFL);
      printf (_("Enter the new value, or press return for the default.\n"));
      new_shell = get_value (pw_data->pw.pw_shell, _("Login Shell"));
      signal (SIGINT, SIG_IGN);
    }

  /* we don't need to change the shell if here is no change */
  if (new_shell == NULL || strcmp (pw_data->pw.pw_shell, new_shell) == 0)
    {
      if (!silent)
	printf (_("Shell not changed.\n"));
      return 0;
    }

  pw_data->new_shell = new_shell;

  if (check_shell (program, new_shell) != 0)
    {
      free_user_t (pw_data);
      return 1;
    }

  if (write_user_data (pw_data, 0) != 0)
    {
      fprintf (stderr, _("Error while changing login shell.\n"));
      free_user_t (pw_data);
      return 1;
    }
  else
    {
#ifdef HAVE_NSCD_FLUSH_CACHE
      nscd_flush_cache ("passwd");
#endif
      if (!silent)
	printf (_("Shell changed.\n"));
    }

  free_user_t (pw_data);

  return 0;
}
